src/binary/Binary.ts

```diff
@@ -0,0 +1,113 @@
+import { Mutex } from "await-semaphore";
+import * as child_process from "child_process";
+import { createInterface, ReadLine } from "readline";
+import {
+  API_VERSION,
+  CONSECUTIVE_RESTART_THRESHOLD,
+  restartBackoff,
+} from "../consts";
+import { sleep } from "../utils";
+import { runTabNine } from "./run";
+
+export default class Binary {
+  private consecutiveRestarts: number = 0;
+  private isRestarting: boolean = false;
+  private mutex: Mutex = new Mutex();
+
+  private proc?: child_process.ChildProcess = null;
+  private rl: ReadLine;
+
+  constructor() {
+    setImmediate(() => {
```

Comment:
why to do async stuff in the ctor?

Reply:
Well originally to not block the initialisation. But I removed it.